### PR TITLE
implement type-hints for all (some) our functions

### PR DESCRIPTION
**Pull request by igalic**
gradually add types on all functions and variables.
this effort is guided by mypy.

### CODE REVIEW

File: libiocage/lib/Distribution.py [hunk 1]
```diff
@@ -27,7 +27,7 @@ class DistributionGenerator:
     def __init__(self, host, zfs=None, logger=None):
         self.logger = libiocage.lib.helpers.init_logger(logger)
         self.zfs = libiocage.lib.helpers.init_zfs(zfs)
-        libiocage.lib.helpers.init_host(self, host)
+        self.host = libiocage.lib.helpers.init_host(host, self.logger)
```
Comment (gronke) [inline libiocage/lib/Distribution.py]: I prefer to use named arguments - at least for logger and those kind of dependencies. It's much easier to recognize broken interfaces when being explicit.

Comment (igalic) [reply libiocage/lib/Distribution.py]: ack, will change

File: libiocage/lib/events.py [hunk 2]
```diff
@@ -15,7 +17,7 @@ class IocageEvent:
     Base class for all other iocage events
     """
 
-    HISTORY = []
+    HISTORY: List[Any] = []
```
Comment (igalic) [inline libiocage/lib/events.py]: this is not ideal, but i have idea how to tell it the type of this class

File: libiocage/lib/helpers.py [hunk 3]
```diff
@@ -1,38 +1,31 @@
...
+        return get_zfs()
 
 
 def get_zfs():
     return libzfs.ZFS(history=True, history_prefix="<iocage>")
 
 
-def init_host(self, host=None):
+def init_host(host: libiocage.lib.Host.Host=None, logger:
+              libiocage.lib.Logger.Logger=None) -> libiocage.lib.Host.Host:
```
Comment (igalic) [inline libiocage/lib/helpers.py]: this return type is now incompatible

File: libiocage/lib/helpers.py [hunk 4]
```diff
@@ -1,38 +1,31 @@
...
-            logger = self.logger
-        except:
-            logger = None
+        return host
 
-        try:
-            self.host = self._class_host(logger=logger)
-        except:
-            self.host = libiocage.lib.Host.HostGenerator(logger=logger)
+    return libiocage.lib.Host.HostGenerator(logger=logger)
```
Comment (igalic) [inline libiocage/lib/helpers.py]: with this return

File: libiocage/lib/events.py (hunk 2 above)
Comment (gronke) [reply libiocage/lib/events.py]: Guess I can explain that: `List[libiocage.lib.events.IocageEvent]`

The class attribute, not the instance attribute contains a list of all events occurred. This way every event has access to the full history.

File: libiocage/lib/helpers.py (hunk 3 above)
Comment (gronke) [reply libiocage/lib/helpers.py]: We're returning a HostGenerator. Internally we don't use the wrappers that flatten the generator to a result list.

File: libiocage/lib/Distribution.py [hunk 5]
```diff
@@ -25,12 +25,11 @@ class DistributionGenerator:
     mirror_link_pattern = r"a href=\"([A-z0-9\-_\.]+)/\""
 
     def __init__(self, host, zfs=None, logger=None):
-        libiocage.lib.helpers.init_logger(self, logger)
-        libiocage.lib.helpers.init_zfs(self, zfs)
-        libiocage.lib.helpers.init_host(self, host)
+        self.logger = libiocage.lib.helpers.init_logger(logger)
+        self.zfs = libiocage.lib.helpers.init_zfs(zfs)
+        self.host = libiocage.lib.helpers.init_host(host, self.logger)
```
Comment (gronke) [inline libiocage/lib/Distribution.py]: Please use named attributes for everything, but the first argument.

File: libiocage/lib/Jails.py [hunk 6]
```diff
@@ -23,9 +23,11 @@ def __init__(self,
                  logger=None,
                  zfs=None):
 
-        libiocage.lib.helpers.init_logger(self, logger)
-        libiocage.lib.helpers.init_zfs(self, zfs)
-        libiocage.lib.helpers.init_host(self, host)
+        self.logger = libiocage.lib.helpers.init_logger(logger)
+        self.host = libiocage.lib.helpers.init_host(host, self.logger)
+
+        # this could be wrong..er.. now?:
```
Comment (gronke) [inline libiocage/lib/Jails.py]: What's this comment for? Do you suspect something to be wrong with the following lines?

File: libiocage/lib/errors.py [hunk 7]
```diff
@@ -388,7 +388,8 @@ def __init__(self, *args, **kwargs):
 
 class MissingFeature(IocageException, NotImplementedError):
 
-    def __init__(self, feature_name: str, plural: bool=False, *args, **kwargs):
+    def __init__(self, feature_name: str, plural: bool=False,
+                 *args, **kwargs) -> None:
```
Comment (gronke) [inline libiocage/lib/errors.py]: Please break every argument into a new line. Like such:

```python
def __init__(
    self
    feature_name: str
    plural: bool=False,
    *args
    **kwargs
) -> None:
    pass
```

File: libiocage/lib/helpers.py (hunk 3 above)
Comment (gronke) [reply libiocage/lib/helpers.py]: `HostGenerator` should be the type returned.

Comment (igalic) [reply libiocage/lib/helpers.py]: ack, thanks

File: libiocage/lib/Jails.py (hunk 6 above)
Comment (igalic) [reply libiocage/lib/Jails.py]: so, we init_zfs(), and then we run the equivalent of the get_zfs() function, and if that's meant to be happening consistently, we may wanna just always do that in init_zfs()

File: libiocage/lib/errors.py (hunk 7 above)
Comment (igalic) [reply libiocage/lib/errors.py]: yapf doesn't work for python 3.6, and standard indent can't seem to be able to read minds 😛

Comment (gronke) [reply libiocage/lib/errors.py]: I'm running this syntax on python3.6. Why should it not work?

File: libiocage/lib/events.py (hunk 2 above)
Comment (igalic) [reply libiocage/lib/events.py]: python (or mypy) can't find the type `libiocage.lib.events.IocageEvent`

Comment (gronke) [reply libiocage/lib/events.py]: Within this file it's just `IocageEvent`, not the fqdn :trollface:

Comment (igalic) [reply libiocage/lib/events.py]: yeah, that's what I thought too, but that doesn't work either.

File: libiocage/lib/errors.py (hunk 7 above)
Comment (igalic) [reply libiocage/lib/errors.py]: done.

File: libiocage/lib/Distribution.py [hunk 8]
```diff
@@ -36,7 +36,7 @@
 
 class DistributionGenerator:
 
-    _class_release = libiocage.lib.Release.ReleaseGenerator
+    _class_release = 'libiocage.lib.Release.ReleaseGenerator'
```
Comment (gronke) [inline libiocage/lib/Distribution.py]: This should not be a string. Later on we instantiate this class.

File: libiocage/lib/Host.py [hunk 9]
```diff
@@ -91,4 +89,4 @@ def processor(self):
 
 class Host(HostGenerator):
 
-    class_distribution = libiocage.lib.Distribution.DistributionGenerator
+    class_distribution = 'libiocage.lib.Distribution.DistributionGenerator'
```
Comment (gronke) [inline libiocage/lib/Host.py]: it should have been `libiocage.lib.Distribution.Distribution` here. Line 35 should not be deleted as well.

File: libiocage/lib/Host.py [hunk 10]
```diff
@@ -32,18 +32,16 @@
...
-        libiocage.lib.helpers.init_zfs(self, zfs)
+        self.logger = libiocage.lib.helpers.init_logger(logger)
+        self.zfs = libiocage.lib.helpers.init_zfs(zfs)
         self.datasets = libiocage.lib.Datasets.Datasets(
             root=root_dataset,
             logger=self.logger,
             zfs=self.zfs
         )
-        self.distribution = self._class_distribution(
+        self.distribution = libiocage.lib.Distribution.DistributionGenerator(
```
Comment (gronke) [inline libiocage/lib/Host.py]: It is important to instantiate from the class attribute _class_distribution. If you use a non-generator class like Host, you don't expect a generator distribution.

File: libiocage/lib/JailConfig.py [hunk 11]
```diff
@@ -79,8 +79,7 @@ def __init__(self,
 
         dict.__init__(self)
 
-        libiocage.lib.helpers.init_logger(self, logger)
-
+        self.logger = libiocage.lib.helpers.init_logger(logger)
```
Comment (gronke) [inline libiocage/lib/JailConfig.py]: I like to use the output from the helper to store explicitly as self.logger, but we should remain passing self as first attribute. This way we do not need to pass other named arguments explicitly later on.

File: libiocage/lib/Jails.py [hunk 12]
```diff
@@ -46,9 +46,11 @@ def __init__(self,
                  logger=None,
                  zfs=None):
 
-        libiocage.lib.helpers.init_logger(self, logger)
-        libiocage.lib.helpers.init_zfs(self, zfs)
-        libiocage.lib.helpers.init_host(self, host)
+        self.logger = libiocage.lib.helpers.init_logger(logger)
+        self.host = libiocage.lib.helpers.init_host(host, logger=self.logger)
```
Comment (gronke) [inline libiocage/lib/Jails.py]: Here's one example why we should pass self as first argument to the init_ helpers.

File: libiocage/lib/errors.py [hunk 13]
```diff
@@ -412,7 +412,11 @@ def __init__(self, *args, **kwargs):
 
 class MissingFeature(IocageException, NotImplementedError):
 
-    def __init__(self, feature_name: str, plural: bool=False, *args, **kwargs):
+    def __init__(self,
+                 feature_name: str,
+                 plural: bool=False,
+                 *args,
+                 **kwargs) -> None:
```
Comment (gronke) [inline libiocage/lib/errors.py]: Please use this style for multi-line method interfaces
```python
def __init__(self,
    feature_name: str,
    plural: bool=False,
    *args,
    **kwargs
) -> None:
```

File: libiocage/lib/Host.py (hunk 10 above)
Comment (gronke) [reply libiocage/lib/Host.py]: @igalic as we discussed offline, this change should be reverted

File: libiocage/lib/Host.py (hunk 9 above)
Comment (igalic) [reply libiocage/lib/Host.py]: n.b.: In `class HostGenerator` this property was `_class_distribution` (yay ☭) but here it's without the leading `_`.

File: libiocage/lib/Host.py [hunk 14]
```diff
@@ -91,4 +91,4 @@ def processor(self):
 
 class Host(HostGenerator):
 
-    class_distribution = libiocage.lib.Distribution.DistributionGenerator
+    _class_distribution = libiocage.lib.Distribution.DistributionGenerator
```
Comment (igalic) [inline libiocage/lib/Host.py]: is this supposed to be a generator?
is it supposed to be `class_distribution` and not `_class_distribution`

Comment (gronke) [reply libiocage/lib/Host.py]: As rule of thumb: when the class is not a generator it is not supposed to expose or use generators at all. So, this should be a Distribution, not DistributionGenerator.

**Comment by gronke** [comment]
@igalic besides that `init_*` functions should remain having `self` as first argument, so that we don't need to pass other attributes (like logger, host, etc) on every helpers.init_* call. It's easier to update on inter-dependency changes in future updates as well.

**Pull request merged**